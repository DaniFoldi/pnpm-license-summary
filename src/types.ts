export type Package = { name: string; version: string  }
export type PackageDetails = Package & { license: string; homepage: string }

export type AllowedLicenses = Set<string>
export type IgnoredPackages = Set<Package>

export type Result = { unusedIgnores: Array<string>; licensesUsed: Record<string, Set<PackageDetails>> }
  & ({ success: true } | { success: false; invalidLicenses: Set<string> })
