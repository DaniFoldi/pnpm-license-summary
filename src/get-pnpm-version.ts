import { spawn } from 'node:child_process'


export async function getPnpmVersion(): Promise<string> {
  const proc = spawn('pnpm', [ '--version' ])

  let stdout = ''
  proc.stdout.on('data', data => stdout += data.toString())

  // eslint-disable-next-line promise/avoid-new
  await new Promise((resolve, reject) => {
    proc.on('close', resolve)
    proc.on('error', reject)
  })

  return stdout.trim()
}
