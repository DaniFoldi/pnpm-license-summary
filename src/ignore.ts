export function matchesIgnore(target: string, template: string): boolean {
  if (template === '*') return true
  if (template === target) return true
  if (template.endsWith('*')) {
    const prefix = template.slice(0, -1)
    return target.startsWith(prefix)
  }
  return false
}
