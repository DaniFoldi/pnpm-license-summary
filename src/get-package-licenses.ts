import { spawn } from 'node:child_process'
import { resolve } from 'node:path'
import { cwd } from 'node:process'
import type { PackageDetails } from './types'
import { getPnpmVersion } from './get-pnpm-version'


async function runPnpmLicenses(directory: string): Promise<string> {
  const proc = spawn('pnpm', [ 'licenses', 'ls', '--json' ], {
    cwd: resolve(cwd(), directory)
  })

  let stdout = ''
  proc.stdout.on('data', data => stdout += data.toString())

  // eslint-disable-next-line promise/avoid-new
  await new Promise((resolve, reject) => {
    proc.on('close', resolve)
    proc.on('error', reject)
  })

  return stdout
}


/**
 *
 * @param {string} directory The directory to get licenses for, needs to contain package.json
 * @returns {Promise<Record<string, Array<PackageDetails>>>} A map of licenses to list of packages
 */
export async function getLicenses(directory: string): Promise<Record<string, Array<PackageDetails>>> {

  const major = (await getPnpmVersion()).split('.', 1).join('')

  switch (major) {
    case '10':
    case '9': {
      const result = await runPnpmLicenses(directory)
      const parsedResult = result.startsWith('{') ? JSON.parse(result) : {}
      return Object
        .entries(parsedResult)
        .map(([ license, packages ]) => {
          return (packages as Array<Omit<PackageDetails, 'version' | 'path'> & { versions: string[]; paths: string[] }>).map(pkg => {
            const { paths, versions, ...rest } = pkg
            return versions.map((version, i) => [license, { version, path: paths[i], ...rest }])
          }).flat()
        })
        .flat()
        .reduce(function(acc, [license, pkg]) {
          // @ts-expect-error shut up
          acc[license] = acc[license] || []
          // @ts-expect-error shut up
          acc[license].push(pkg)
          return acc
          }, {}
        );
    }
    case '8': {
      const result = await runPnpmLicenses(directory)
      return result.startsWith('{') ? JSON.parse(result) : {}
    }
    default:
      throw new Error('Unsupported pnpm version, please use pnpm 8 or 9.')
  }
}
