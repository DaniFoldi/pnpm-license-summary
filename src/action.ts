import { debug } from '@actions/core'
import { getLicenses } from './get-package-licenses'
import { matchesIgnore } from './ignore'
import type { AllowedLicenses, IgnoredPackages, Result } from './types'

/**
 *
 * @param {string} directory The directory to get licenses for, needs to contain package.json
 * @param {AllowedLicenses} allowedLicenses Set of allowed licenses
 * @param {IgnoredPackages} ignoredPackages Set of 'package @ version' to ignore
 * @returns {Promise<Result>} The result of the action
 */
export async function action(
  directory: string,
  allowedLicenses: AllowedLicenses,
  ignoredPackages: IgnoredPackages
): Promise<Result> {
  let result: Result = {
    success: true,
    unusedIgnores: [],
    licensesUsed: {}
  }

  const licenses = await getLicenses(directory)

  ignores: for (const ignored of ignoredPackages) {
    for (const [ _, packages ] of Object.entries(licenses)) {
      if (!packages || !Array.isArray(packages)) {
        debug(`Something went wrong with ${ignored.name}@${ignored.version}`)
        debug(`Licenses: ${JSON.stringify(licenses, null, 2)}`)
        continue
      }
      if (packages.some(pkg => matchesIgnore(pkg.name, ignored.name) && matchesIgnore(pkg.version, ignored.version))) {
        continue ignores
      }
    }
    result.unusedIgnores.push(`${ignored.name}@${ignored.version}`)
  }

  for (const [ license, packages ] of Object.entries(licenses)) {
    if (!Array.isArray(packages)) {
      debug(`No packages found using ${license}`)
      continue
    }
    licenses[license] = packages
      .filter(pkg => {
        for (const { name, version } of ignoredPackages) {
          if (matchesIgnore(pkg.name, name) && matchesIgnore(pkg.version, version)) {
            return false
          }
        }
        return true
      })
  }

  result.licensesUsed = Object.fromEntries(Object.entries(licenses)
    .filter(([ _, packages ]) => Array.isArray(packages))
    .map(([ license, packages ]) => ([ license, new Set(packages) ] as const))
    .sort((a, b) => b[1].size - a[1].size))

  const invalidLicenses = new Set(Object.entries(licenses)
    .filter(([ license ]) => !allowedLicenses.has(license))
    .filter(([ _, packages ]) => packages.length > 0)
    .map(([ license ]) => license))

  if (invalidLicenses.size > 0) {
    result = {
      ...result,
      invalidLicenses,
      success: false
    }
  }

  return result
}
