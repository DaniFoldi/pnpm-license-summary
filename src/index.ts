import { setFailed, getMultilineInput, getInput, warning, info } from '@actions/core'
import { Table } from 'console-table-printer'
import { action } from './action'


try {
  const directory = (getInput('directory', { required: false, trimWhitespace: true }) || '.').replaceAll('\'', '')
  const allowed_licenses = new Set(getMultilineInput('allowed', { required: false }) ?? [])
  const ignored_packages = new Set(getMultilineInput('ignored', { required: false }).map(pkg => {
    if (pkg.startsWith('@')) {
      const [ _, name, version ] = pkg.split('@')
      return { name: `@${name}`, version }
    }
    const [ name, version ] = pkg.split('@')
    return { name, version }
  }))

  const result = await action(directory, allowed_licenses, ignored_packages)

  if (result.success === true) {
    info('All licenses are valid')
    const table = new Table({
      columns: [
        { alignment: 'left', name: 'license', title: 'License' },
        { alignment: 'right', name: 'count', title: 'Count' }
      ]
    })
    table.addRows(Object.entries(result.licensesUsed).map(pkg => ({
      license: pkg[0],
      count: pkg[1].size
    })))
    info(table.render())
  } else {
    setFailed('Invalid licenses found')
    for (const license of result.invalidLicenses) {
      warning(`Invalid license ${license}`)
      const table = new Table({
        columns: [
          { alignment: 'left', name: 'name', title: 'Name' },
          { alignment: 'left', name: 'version', title: 'Version' },
          { alignment: 'left', name: 'license', title: 'License' },
          { alignment: 'left', name: 'homepage', title: 'Homepage' }
        ]
      })
      table.addRows([ ...result.licensesUsed[license] ].map(pkg => ({
        name: pkg.name,
        version: pkg.version,
        license: pkg.license,
        homepage: pkg.homepage
        // todo include chain `pnpm why <package> --json`
      })))
      info(table.render())
    }
  }

} catch (error) {
  setFailed(error.message)
}
