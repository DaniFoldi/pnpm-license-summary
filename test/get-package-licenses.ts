import { expect, it } from 'vitest'
import { getLicenses } from '../src/get-package-licenses'


it('returns an empty object on empty', async () => {
  const licenses = await getLicenses('fixtures/empty')
  expect(Object.entries(licenses)).toHaveLength(0)
})

it('returns correct object of packages', async () => {
  const licenses = await getLicenses('.')
  expect(licenses).matchSnapshot()
})
