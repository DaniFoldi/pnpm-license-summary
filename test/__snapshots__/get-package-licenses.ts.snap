// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns correct object of packages 1`] = `
{
  "error": {
    "code": "ERR_PNPM_MISSING_PACKAGE_INDEX_FILE",
    "message": "Failed to find package index file for /@actions/core/1.10.1, please consider running 'pnpm install'",
  },
}
`;
